NoSuchElementException occurs when execute `PREVIEW FETCH 2 FROM "t_order_cursor"`

## Bug Report

### Which version of ShardingSphere did you use?

https://github.com/apache/shardingsphere/commit/72b683cc3d431452b0e645f5bfcfa2f9480e65f2

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-Proxy

### Expected behavior

Show preview result of `PREVIEW FETCH 2 FROM "t_order_cursor"` statement.

### Actual behavior

```
[ERROR] 2022-06-08 10:09:57.017 [Connection-7-ThreadExecutor] o.a.s.p.f.c.CommandExecutorTask - Exception occur: 
java.util.NoSuchElementException: null
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1447)
	at java.util.HashMap$KeyIterator.next(HashMap.java:1469)
	at org.apache.shardingsphere.sharding.route.engine.type.ShardingRouteEngineFactory.getDDLRoutingEngine(ShardingRouteEngineFactory.java:124)
	at org.apache.shardingsphere.sharding.route.engine.type.ShardingRouteEngineFactory.newInstance(ShardingRouteEngineFactory.java:94)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext(ShardingSQLRouter.java:57)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext(ShardingSQLRouter.java:45)
	at org.apache.shardingsphere.infra.route.engine.impl.PartialSQLRouteExecutor.route(PartialSQLRouteExecutor.java:68)
	at org.apache.shardingsphere.infra.route.engine.SQLRouteEngine.route(SQLRouteEngine.java:53)
	at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.route(KernelProcessor.java:54)
	at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.generateExecutionContext(KernelProcessor.java:46)
	at org.apache.shardingsphere.proxy.backend.text.distsql.ral.advanced.PreviewDistSQLBackendHandler.getRows(PreviewDistSQLBackendHandler.java:115)
	at org.apache.shardingsphere.proxy.backend.text.distsql.ral.QueryableRALBackendHandler.handle(QueryableRALBackendHandler.java:49)
	at org.apache.shardingsphere.proxy.backend.text.distsql.ral.RALBackendHandler.execute(RALBackendHandler.java:47)
	at org.apache.shardingsphere.proxy.frontend.opengauss.command.query.simple.OpenGaussComQueryExecutor.execute(OpenGaussComQueryExecutor.java:76)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:107)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[INFO ] 2022-06-08 10:20:25.112 [HikariPool-4 connection adder] o.o.core.v3.ConnectionFactoryImpl - [2bc6ae6a-8910-4009-b353-b2b961cbb53c] Try to connect. IP: localhost:15432
[INFO ] 2022-06-08 10:20:25.137 [HikariPool-4 connection adder] o.o.core.v3.ConnectionFactoryImpl - [127.0.0.1:53892/ocalhost/127.0.0.1:15432] Connection is established. ID: 2bc6ae6a-8910-4009-b353-b2b961cbb53c
```

### Reason analyze (If you can)

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

### Example codes for reproduce this issue (such as a github link).

--- a/shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/data/impl/SchemaAssignedDatabaseBackendHandler.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/data/impl/SchemaAssignedDatabaseBackendHandler.java
@@ -95,13 +95,14 @@ public final class SchemaAssignedDatabaseBackendHandler implements DatabaseBacke
         if (statementContext instanceof CursorStatementContext) {
             connectionSession.getCursorDefinitions().put(cursorName, (CursorStatementContext) statementContext);
         }
-        if (statementContext instanceof CloseStatementContext) {
-            FetchOrderByValueQueuesHolder.get().remove(cursorName);
-        }
         if (statementContext instanceof CursorDefinitionAware) {
             CursorStatementContext cursorStatementContext = connectionSession.getCursorDefinitions().get(cursorName);
             Preconditions.checkArgument(null != cursorStatementContext, "Cursor %s does not exist.", cursorName);
             ((CursorDefinitionAware) statementContext).setUpCursorDefinition(cursorStatementContext);
+        }
+        if (statementContext instanceof CloseStatementContext) {
+            FetchOrderByValueQueuesHolder.get().remove(cursorName);
+            connectionSession.getCursorDefinitions().remove(cursorName);
         }
     }
     

--- a/shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/advanced/PreviewDistSQLBackendHandler.java
+++ b/shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/advanced/PreviewDistSQLBackendHandler.java
@@ -22,7 +22,10 @@ import com.google.common.base.Strings;
 import org.apache.shardingsphere.distsql.parser.statement.ral.advanced.PreviewStatement;
 import org.apache.shardingsphere.infra.binder.LogicSQL;
 import org.apache.shardingsphere.infra.binder.SQLStatementContextFactory;
+import org.apache.shardingsphere.infra.binder.aware.CursorDefinitionAware;
 import org.apache.shardingsphere.infra.binder.statement.SQLStatementContext;
+import org.apache.shardingsphere.infra.binder.statement.ddl.CursorStatementContext;
+import org.apache.shardingsphere.infra.binder.type.CursorAvailable;
 import org.apache.shardingsphere.infra.binder.type.TableAvailable;
 import org.apache.shardingsphere.infra.config.props.ConfigurationPropertyKey;
 import org.apache.shardingsphere.infra.context.kernel.KernelProcessor;
@@ -107,6 +110,9 @@ public final class PreviewDistSQLBackendHandler extends QueryableRALBackendHandl
         if (sqlStatementContext instanceof TableAvailable) {
             ((TableAvailable) sqlStatementContext).getTablesContext().getDatabaseName().ifPresent(SQLStatementDatabaseHolder::set);
         }
+        if (sqlStatementContext instanceof CursorAvailable && sqlStatementContext instanceof CursorDefinitionAware) {
+            setUpCursorDefinition(sqlStatementContext);
+        }
         ShardingSphereDatabase database = ProxyContext.getInstance().getDatabase(connectionSession.getDatabaseName());
         if (!database.isComplete()) {
             throw new RuleNotExistedException();
@@ -117,6 +123,13 @@ public final class PreviewDistSQLBackendHandler extends QueryableRALBackendHandl
                 ? getFederationExecutionUnits(logicSQL, databaseName, metaDataContexts)
                 : executionContext.getExecutionUnits();
         return executionUnits.stream().map(this::buildRow).collect(Collectors.toList());
+    }
+    
+    private void setUpCursorDefinition(final SQLStatementContext<?> sqlStatementContext) {
+        String cursorName = ((CursorAvailable) sqlStatementContext).getCursorName().getIdentifier().getValue().toLowerCase();
+        CursorStatementContext cursorStatementContext = connectionSession.getCursorDefinitions().get(cursorName);
+        Preconditions.checkArgument(null != cursorStatementContext, "Cursor %s does not exist.", cursorName);
+        ((CursorDefinitionAware) sqlStatementContext).setUpCursorDefinition(cursorStatementContext);
     }
     
     private List<Object> buildRow(final ExecutionUnit unit) {

